### ISSUE TITLE
推理开销对比:deeplob vs binctabl(参数量 / 单样本延迟)

### ISSUE THREAD

**Issue by liningping** [ningpingli/hft_lob#16]
# 推理开销对比:deeplob vs binctabl(参数量 / 单样本延迟)

## 动机

补模型选型缺的一个维度:同一数据契约、同一实现路径下,deeplob 与 binctabl 的**参数量**和
**单样本推理延迟**(微秒)差多少。不改模型代码,只测量。

## 方法

- 数据契约:不可变包 `a052b81a…`(688981)的 `dataset.json` —— `num_features=23`、
  `history_snapshots=180`、`levels=5`(ASKp1..ASKp5)、`labels=[60]` → `target_count=1`。
- 模型实例化走训练/测试同一条工厂 `hft_lob.models.lob_model.build_model` + `configs/models/<model>.yaml`
  (非手写等价结构),`torch.manual_seed(config.seed)`。
- 环境:RTX 4090(空闲,index 0)、torch 2.13.0+cu130 / CUDA 13.0、float32、`model.eval()` +
  `torch.inference_mode()`;未开 `cudnn.benchmark`(与 `utils/seed.py` 的训练默认一致)。
- 计时:同一 batch 张量连续前向 `iters` 次,CUDA event 夹住整段后除以次数 → μs/batch;
  每档 20 个 run 取均值/标准差,20 个 run 的极差即 std 列。warmup = iters/5。
  batch=1: 300 iter/run;batch=128: 100 iter/run。
  **单样本延迟 = μs/batch ÷ batch**,即下表 μs/sample。
- 交叉校验:wall-clock 每迭代 `synchronize` 口径 + `torch.profiler` 自我 CUDA 时间 + kernel 计数
  (见"归因"),确认聚合计时不是假象。

## 结果

| 模型 | 参数量 | GPU μs/sample @B=1 | GPU μs/sample @B=128 | CPU 1 线程 μs/sample @B=1 | CPU 1 线程 @B=128 |
|---|---|---|---|---|---|
| deeplob | 138,657 | **1064.9** (std 19.4) | **10.16** (std 15.0) | 2541.9 | 1476.0 |
| binctabl | 19,845 | **807.9** (std 1.5) | **6.53** (std 2.1) | 427.9 | 157.3 |
| 比值 deeplob/binctabl | **6.99×** | 1.32× | 1.56× | 5.94× | 9.38× |

两个模型都是全部参数可训练。binctabl 参数量手算可复核:BiN 408 + BL1 4,260 + BL2 15,025 + TABL 152 = **19,845** ✓。

## 归因:B=1 由 kernel 数 / 启动开销主导,不是算力

1. **profiler(单次 B=1 前向)**:deeplob 89 个 CUDA kernel、自我 CUDA 时间 238.9 μs;
   binctabl 82 个 kernel、153.6 μs。即实测延迟里 GPU 真正忙的比例只有 **22% / 19%**。
2. **环境 kernel 启动开销标定**(堆叠 `Linear` 探针):2 / 14 / 29 / 59 个 kernel →
   19.8 / 123.2 / 247.0 / 495.1 μs,线性,**≈8.4-9.9 μs/kernel**。
   按 8.4 μs 推算 89 kernel ≈ 0.75 ms、82 kernel ≈ 0.69 ms,与实测同量级;
   余量为真实计算与布局转换——deeplob 单次前向含 16 次 `cudnn::nchwToNhwcKernel`,
   其 conv 路径几乎全是小 kernel。
3. **binctabl 的 `_enforce_max_norm` 是纯开销**:该函数在 `BiN_CTABL.forward` 内被调用 **7 次**,
   每次 `torch.linalg.matrix_norm(w) > 10.0` 都触发 device→CPU 同步(profiler 可见 7 次 `Memcpy DtoH`)。
   把该调用打桩为 no-op 后:
   - B=1: 807.9 → **577.5 μs/sample**(−230.4 μs,**−28.5%**)
   - B=128: 6.53 → **4.73 μs/sample**(−27.6%)

   单独计时这 7 次调用 = 202.6 μs(B=1)。本次随机初始化下所有权重矩阵 norm 均 < 10,
   即分支从未真的钳制——这 230 μs 全是**检查**的成本。
4. **wall-clock(每迭代 synchronize)复核**:deeplob 1068.6 μs、binctabl 1034.6 μs。
   deeplob 与聚合口径(1064.9)一致,确认其测量无假象;binctabl 偏高是因为 forward 内含 7 次
   阻塞式 DtoH,主机无法提前排空队列,GPU 空转门到门时被放大。
   稳态单流(部署口径)= 表中的聚合值 807.9 μs;两者的真实区间是 0.81-1.03 ms,不改变结论排序。

## 结论

1. **参数量**:binctabl 比 deeplob 小 **7 倍**(19,845 vs 138,657)。
2. **单样本推理延迟(B=1, RTX 4090, fp32 eager)**:deeplob **1.06 ms**,binctabl **0.81 ms**,
   deeplob 慢 **1.32×**。差距远小于参数量差距,因为两者都被 kernel 数 × 启动开销主导。
3. **批量摊销(B=128,评测口径)**:deeplob 10.16 μs/sample,binctabl 6.53 μs/sample,deeplob 慢 1.56×。
4. **CPU 单线程**差距放大到 5.94×(B=1)/ 9.38×(B=128):binctabl 矩阵规模小、无 LSTM 时间步循环,
   deeplob 的 conv/LSTM 在 CPU 上是最贵的一档。
5. 结合 IC(报告 §2.3/§2.4,v3 fold 50-59:deeplob mean IC 0.0620 / 有效率 200/200;
   binctabl 0.0956 / 123/200),**binctabl 在开销与 IC 两个维度都不吃亏**;
   但其高 IC 依赖 fold_050-055,fold_056-059 崩(训练稳定性问题,不在本 issue 范围)。

## 建议(未执行)

1. **去掉 forward 内每次调用的 `_enforce_max_norm`**(binctabl 单样本延迟 −28%):该约束是训练
   稳定性手段,应挪到 optimizer step 之后或每 K 步执行一次,而不是 forward 内 7 次带主机同步的检查。
   改动局部、收益明确,需要时单开改动。
2. B=1 的 ~0.8-1.1 ms 是 eager 启动开销主导;若日后出现延迟约束,`torch.compile` / CUDA Graph
   是下一档杠杆(本次未测)。
3. 当前 60s horizon 预测下 ms 量级不构成约束——本 issue 的价值在于**选型时不必为 binctabl 的
   大矩阵担心延迟**,以及 `_enforce_max_norm` 的固定开销已被量化。

## 产物 / 复现

- 主脚本:`output/run_logs/bench_inference_speed.py`(表格 + JSON)
- 归因脚本:`output/run_logs/bench_profile.py`(profiler + 分段计时)
- 原始数据:`output/run_logs/bench_inference_speed.json`
- 复现:
  `uv run python output/run_logs/bench_inference_speed.py --gpu-id 0 --runs-cuda 20 --iters-cuda 300 --iters-cuda-big 100`
- 报告摘要:`output/EXPERIMENT_REPORT.md` 末节「推理开销对比:deeplob vs binctabl(2026-09-11)」

## 边界

- 只测前向,不含数据加载 / 预处理 / 后处理;输入张量常驻显存。
- 单卡 RTX 4090、fp32、eager、未编译;换 GPU / 精度 / batch 会改变绝对值与倍数。
- 图中权重取初始化值(norm 检查分支为 no-op);延迟与权重取值无关,但若训练后某 `W` 的 norm > 10,
  `_enforce_max_norm` 会真的做一次 in-place 更新,开销同量级。
- B=128 是吞吐口径(per-sample 摊销),不是单请求延迟;未测多请求排队场景。

**Comment by liningping** [comment]
## 补充:参数量少 7×,为什么延迟只快 1.32×?(2026-09-11 追测)

### TL;DR

两个原因叠加:

1. **参数量 ≠ 计算量**:实测算术量差 **61.7×**(43.79M vs 0.71M MACs),不是 7×。
2. **B=1 时两者的瓶颈都是 kernel 派发固定开销,不是算力**:GPU 利用率只有 22% / 19%,
   端到端延迟里 ~70% 是与模型大小无关的固定成本。
   算术少 61.7× 只换来 GPU 忙时间少 1.55×、端到端少 1.32×。

### 1. 实测算术量:43.79M vs 0.71M MACs

MACs 用 `torch.utils.flop_counter.FlopCounterMode` 实测(非估算):

| 模型 | 参数量 | MACs @B=1 | MACs @B=128 | MACs / param |
|---|---|---|---|---|
| deeplob | 138,657 | 43.79 M | 5,604.6 M | 316 |
| binctabl | 19,845 | 0.71 M | 89.2 M | 36 |
| 比值 | 6.99× | **61.7×** | **62.8×** | 8.8× |

参数量为什么低估了算术差距:

- **binctabl 参数高度集中**:`BL2.W1` 是 120×120 = 14,400 个参数 = 全部参数的 **72.6%**,
  但它作用在 `[1,120,5]` 上(时间维已被压到 5)→ 每样本只贡献 **72 k MACs**。
- **deeplob 参数被时间轴大量复用**:conv(72.9 k 参数)在每个空间位置复用;
  `LSTM`(65.8 k 参数)在 **162 个时间步**上各用一次 → 10.6 M MACs。参数利用率高 8.8×。
- 根因是**压缩时间的时机**:binctabl 第二层就把 180 步压成 5 步(再压成 1 步),之后的矩阵再大
  也只算 5 个时间片;deeplob 全程保持时间分辨率(180 → 174 → 168 → 162)。

### 2. 瓶颈不在算术:GPU 利用率 <25%

| batch | 模型 | CUDA op 数/前向 | GPU 忙时间/前向 | 每 op GPU 时间 | 端到端延迟 /sample | 实测吞吐 | 占 4090 fp32 峰值 |
|---|---|---|---|---|---|---|---|
| 1 | deeplob | 89 | 238.7 μs | 2.68 μs | **1064.9 μs** | 0.37 TFLOP/s | 0.4% |
| 1 | binctabl | 82 | 154.0 μs | 1.88 μs | **807.9 μs** | 0.009 TFLOP/s | 0.01% |
| 128 | deeplob | 89 | 1341.0 μs | 15.07 μs | 10.16 μs | 8.36 TFLOP/s | 10% |
| 128 | binctabl | 83 | 256.2 μs | 3.09 μs | 6.53 μs | 0.70 TFLOP/s | 0.85% |

(4090 fp32 峰值 ≈ 82.6 TFLOP/s = 128 SM × 128 lane × 2 FLOP × 2.52 GHz。)

### 3. 257 μs 差额的精确构成(B=1)

`1064.9 − 807.9 = 257.0 μs`,拆成两部分:

| 项 | deeplob | binctabl | 差 |
|---|---|---|---|
| GPU 忙时间 | 238.7 μs | 154.0 μs | **+84.7 μs** |
| 未被 GPU 覆盖的开销(派发 / 同步 / Python)= 端到端 − GPU 忙 | 826.2 μs | 653.9 μs | **+172.3 μs** |
| 合计 | | | **+257.0 μs** ✓ |

- binctabl 的 653.9 μs 里有 **230 μs 是 7 次 `_enforce_max_norm` 的 DtoH 同步**(见上一条归因);
  减掉后其"纯派发"约 424 μs / 82 op。
- deeplob 的 89 个 op 里含 **16 次 `cudnn::nchwToNhwcKernel`**、**14 次 BatchNorm**,cudnn conv 的单次
  派发(descriptor / heuristic / 布局转换)比 gemm 贵 —— 所以它不止多 7 个 op,单位派发成本也更高。
- 参考下限:裸 `Linear` 堆叠探针标定 **≈8.4-9.9 μs/kernel**(与模型规模无关);
  89 kernel × 8.4 ≈ 0.75 ms,已占 deeplob 端到端延迟的 **70%**。
  **快不快看 kernel 数,不看参数量** —— 两个模型的 CUDA op 数只差 8%(89 vs 82)。

### 4. B=128 一档同样如此:算术差 61.7× → GPU 忙时间只差 5.23×

deeplob 已进入 GPU 受限(忙 1341 μs ≈ 端到端 1301 μs),binctabl 仍处于派发受限
(忙 256 μs vs 端到端 836 μs)。即使 B=128、530 k 个激活值,binctabl 也只跑到 4090 fp32 的 **0.85%**——
**它太小,填不满 GPU**,所以"小"在算术层面根本兑现不成速度。

### 5. 对选型的修正

- 想拿到与"算术少 61.7×"相称的收益,必须先消掉固定开销:`_enforce_max_norm` 移出 forward(−28%,见上)、
  `torch.compile` / CUDA Graph 合并 80+ 次派发(未测)。否则 binctabl 在 B=1 的天花板也就是
  "把 82 次派发砍到很少"。
- 反过来看 deeplob:它慢的 1.32× 里,只有 85 μs 是"真算得多",其余 172 μs 是"op 更贵/更难派发"。
- 口径提醒:延迟敏感的单样本场景 1064.9 vs 807.9 μs 差别不大;大 batch(评测/回测)口径
  10.16 vs 6.53 μs/sample(1.56×);CPU 单线程才体现 5.94× / 9.38×。

### 复现

```bash
uv run python output/run_logs/bench_flops_probe.py --gpu-id 0   # 实测 MACs + op 数 + GPU 忙时间
uv run python output/run_logs/bench_inference_speed.py --gpu-id 0 --runs-cuda 20 --iters-cuda 300
```

**Comment by liningping** [comment]
## 已实施:把 `_enforce_max_norm` 移出 forward(issue「建议 1」)

### 改动

- **`src/hft_lob/models/TABL/bin_tabl.py`**
  - `forward` 内 5 次(BTABL)/ 7 次(CTABL)`_enforce_max_norm` 调用全部移除 —— 每次调用都会
    触发 device→CPU 同步。
  - 新增 `enforce_weight_constraints(model)`:遍历 BL / TABL 层,对 `W1/W2/W` 施加 ‖·‖ ≤ 10。
  - 两个模型 `__init__` 末尾调用一次。**必要**:旧实现在第一次前向内钳制,而初始化后
    `BL.W1` 的范数实测 ≈15.5 —— 若构造期不钳而只在步后钳,第一步前向会用到超界权重,
    训练轨迹会偏(此即下面 A/B 中 `old` 变体必须关掉构造期钳制的原因)。
- **`src/hft_lob/trainner/lob_module.py`**
  - 覆写 `optimizer_step(...)`:先 `super()`,再 `enforce_weight_constraints(self.model)`。
    Lightning 该钩子在 `accumulate_grad_batches != 1` 时只在真正 step 的那次调用,
    故每个优化器更新恰好投影一次。
- 未动:`BiN.forward` 的 `y1/y2.clamp_`、`TABL_layer.forward` 的 `l.clamp_`
  (原地 clamp,无主机同步,开销可忽略)。

### 验证 1:延迟(改后重测,同一脚本同一机器)

| 模型 | GPU μs/sample @B=1 | @B=128 | CPU 1 线程 @B=1 | @B=128 |
|---|---|---|---|---|
| deeplob(对照) | 1064.9 → **1060.7** | 10.16 → 10.15 | 2541.9 → 2570.5 | 1476.0 → 1455.8 |
| binctabl | 807.9 → **580.0**(**−28.2%**) | 6.53 → **4.73**(−27.6%) | 427.9 → 368.3 | 157.3 → 158.6 |

与打桩预测的 577.5 μs 一致;deeplob 无 TABL 层,不受影响 ⇒ 改动的收益确实来自那 7 次同步检查。

### 验证 2:约束在真实训练循环里成立

fold_050、2 epoch、binctabl、真实数据 + AdamW + Lightning,`exit=0`,
test/loss 1.21e-6、val/mean_daily_ic 0.0671、evaluation.yaml 20 个有效交易日。训练后
checkpoint 里 7 个受约束权重全部在界内,**max ‖W‖ = 9.52**,且逼近边界 ⇒ 投影确实在工作。

### 验证 3:语义等价性 A/B

同一进程、真实数据、真实优化器(AdamW)、200 步,四种变体
(`output/run_logs/ab_constraint_timing.py`):

| 对比 | 参数最大差 | BL2.W1 范数 |
|---|---|---|
| `new` vs `new` 重跑 | **0.0**(逐位相同) | 9.9356 |
| `new` vs `old`(旧实现) | **1.192e-07**(= float32 1 ULP) | 9.9356 |
| `new` vs 完全不钳制(对照) | **5.194e-01** | 15.4813 |

- 同代码两次运行、跨进程 200 步参数 sha256 完全一致 ⇒ 本机训练是确定性的,
  上表差值可归因于改动本身,不是噪声;
- `old` 与 `new` 只差 **1 个 ULP** ⇒ 投影时机不改变"前向看到什么权重",数学等价;
- 对照(完全去掉约束)差 0.52、范数涨到 15.5 ⇒ 实验灵敏度足够,不是"测不出差异"。

**可复现性提示**:训练对 ULP 级差异呈混沌放大 —— 同为 2 epoch 的 fold_050 实跑,
`new` 与 `old` 权重最大差 **0.035**、测试 mean_daily_ic **0.0981 vs 0.0972**(mse 2.4124e-6 vs 2.4124e-6)。
即**既往 binctabl/binbtabl sweep 的结果不能按位复现**;但 0.0009 的 IC 差远小于该模型自身的
fold 间波动(报告 §2.4 binctabl:0.0035–0.1206),结论层面无影响。需要严格对齐旧数字时,
应按新代码重跑该模型。

### 测试

- 新增 7 个用例:`forward` 不再原地改写权重(旧实现会 → 回归防护)、构造即满足约束、
  `enforce_weight_constraints` 覆盖全部 7 个权重矩阵、约束时机等价性(逐位比较,含
  "确实发生了钳制"的非平凡断言,失败于旧实现)、`optimizer_step` 接线。
- 全量 `uv run pytest -q` = **148 passed, 1 failed**;唯一失败是本 issue 范围外的历史问题
  `tests/trainner/test_executor.py`(cnn1,standalone 与 walk-forward 推理浮点 1e-6 差异被严格相等比较,
  报告 §发现的缺陷 #5 已记录)。`uv run ruff check src tests` 通过。

### 产物

- A/B 工具:`output/run_logs/ab_constraint_timing.py`
- 延迟复测:`output/run_logs/bench_inference_speed.py` / `.json`
- 冒烟训练:临时 config 与输出已按惯例删除(数据、指标见上)
